cmd/devicesimulator/main.go

```diff
@@ -153,12 +186,22 @@ func main() {
 	}()
 
 	log.Infoln("running agents")
-	for i := 0; i < *numDevices; i++ {
-		// stagger the start of each agent
-		time.Sleep(time.Duration(rand.Float64() * float64(agentConfigTemplate.StatusUpdateInterval))) //nolint:gosec
-		agent := agents[i]
-		go startAgent(ctx, agent, log, i)
-		go approveAgent(ctx, log, serviceClient, agentsFolders[i], formattedLables)
+	// limit the maximum number of devices that are being approved.
+	sem := semaphore.NewWeighted(int64(*maxConcurrency))
+	for i := range *numDevices {
+		if err := sem.Acquire(ctx, 1); err != nil {
+			break
+		}
+		go func() {
```

Comment:
What is the advantage of starting a goroutine to start other goroutines ?

Reply:
I think the change could lead to bursty behavior.  Assuming the time to enroll is fixed and relatively long and we use the default timeout (2 seconds), so in the first 2 seconds all the goroutines will be spawned.  And then we need to wait for them until they are over.
The previous implementation without the goroutine was better.  We can enhance it with the maxConcurrency.  Also the time to wait between launching the goroutines should not depend on `StatusUpdateInterval` but it should be a separate command line parameter.

Reply:
>  so in the first 2 seconds all the goroutines will be spawned.

This is only true if:
1. maxConcurrency >= count
2. We're actually able to onboard all devices within 2 seconds

Otherwise, the semaphore limits the number of spawned goroutines to 'maxConcurrency'

> The previous implementation without the goroutine was better.

The previous implementation launches a single agent every `StatusUpdateInterval` / 2 (on average). This, in my opinion, is very very slow. 

Agree with StatusUpdateInterval not being used as our jitter interval. I'm assuming someone chose it for simplicity and to align the jitter with the interval that the agent itself will actually be doing work, but I have no issues changing this to a parameter.